packages/darklang/languageTools/lsp-server/fileSystemProvider.dark

```diff
@@ -0,0 +1,828 @@
+module Darklang =
+  module LanguageTools =
+    module LspServer =
+      module FileSystemProvider =
+        type FileType =
+          | File
+          | Directory
+          | SymbolicLink
+
+        type FileSystemErrorType =
+          | FileNotFound
+          | FileExists
+          | FileNotADirectory
+          | FileIsADirectory
+          | NoPermissions
+          | Unavailable
+          | Other
+
+        type FileSystemProviderClientCapabilities = { queryable: Bool }
+
+        type URI = String
+
+        module FileSystemProviderOptions =
+          type FileSystemProviderOptions =
+            { scheme: String
+              caseSensitive: Bool
+              readonly: Bool }
+
+          let toJson (options: FileSystemProviderOptions) : Json =
+            [ Stdlib.Option.Option.Some(("scheme", Json.String options.scheme))
+              Stdlib.Option.Option.Some(
+                ("caseSensitive", Json.Bool options.caseSensitive)
+              )
+              Stdlib.Option.Option.Some(("readonly", Json.Bool options.readonly)) ]
+            |> Stdlib.Option.values
+            |> Json.Object
+
+
+        module ReadFile =
+          let method = "fileSystem/read"
+
+          module ReadFileParams =
+            type ReadFileParams = { uri: URI }
+
+            let fromJson
+              (json: Json)
+              : Stdlib.Result.Result<ReadFileParams, String> =
+              match json with
+              | Object fields ->
+                match
+                  Stdlib.List.findFirst fields (fun (key, _) -> key == "uri")
+                with
+                | Some((_, String uri)) ->
+                  (ReadFileParams { uri = uri }) |> Stdlib.Result.Result.Ok
+                | _ -> Stdlib.Result.Result.Error "Expected uri"
+              | _ -> Stdlib.Result.Result.Error "Expected object"
+
+
+          module ReadFileResponse =
+            type ReadFileResponse = { content: String }
+
+            let toJson (r: ReadFileResponse) : Json =
+              [ Stdlib.Option.Option.Some(("content", Json.String r.content)) ]
+              |> Stdlib.Option.values
+              |> Json.Object
+
+
+          let handleReadFileRequest
```

Comment:
This could use a bit of tidying:
- there are a few redundant branches here -- the lines [see bottom] are repeated throughout many branches. maybe just have a bit of code at the top that is meant to _just_ get the text to return, and then everything feeds into the "actually return it" bit of code?
- CLEANUP although this isn't part of the official LSP spec, these file system provider handshakes could be of use outside of _our_ language server. maybe we can abstract some functionality out to another module? (`Darklang/lsp-extensions/fileSystemProvider` or something?). At least, the JSON types and mappings could be there.

(redundant code:
```
let responseJson =
                  (ReadFileResponse.ReadFileResponse { content = prettyPrinted })
                  |> ReadFileResponse.toJson
                  |> (fun r ->
                    JsonRPC.Response.Ok.make
                      (Stdlib.Option.Option.Some(requestId))
                      r)
                  |> Stdlib.AltJson.format

                logAndSendToClient responseJson

                // send a showDocument request
                let showDocRequestString =
                  ShowDocument.makeShowDocumentRequest params.uri requestId

                logAndSendToClient showDocRequestString
                state
```
)

Reply:
> CLEANUP although this isn't part of the official LSP spec, these file system provider handshakes could be of use outside of our language server. maybe we can abstract some functionality out to another module? (Darklang/lsp-extensions/fileSystemProvider or something?). At least, the JSON types and mappings could be there.

I kept the changes for this ^ in a separate [commit](https://github.com/darklang/dark/pull/5431/commits/06e587ed92e8863174d8448bea90ca3953a91e6). Please let me know if I misunderstood what you meant